### ISSUE TITLE
Code not working with up-to-date PyTorch & Possible Fix

### ISSUE THREAD

**Issue by egaznep** [lorenlugosch/pytorch_HMM#1]
Hi,

As you have mentioned in the Jupyter notebook, the code does not work with up-to-date PyTorch. Today I was investigating that, and now I figured out the problem. Essentially, some in-place assignment operation is taking place and autograd cannot handle the back-propagation. Some debugging yielded the following fix:

`log_domain_matmul` function has the following lines to broadcast input arrays:

`
log_A_expanded = torch.stack([log_A] * p, dim=2);
log_B_expanded = torch.stack([log_B] * m, dim=0)
`
Instead of manually expanding the arrays, I just reshaped them and let pytorch handle the broadcasting by doing the following:

`        
log_A_expanded = torch.reshape(log_A, (m,n,1));
log_B_expanded = torch.reshape(log_B, (1,n,p))
`

Right now I am training a model with pytorch 1.7.1 (AFAIK the latest). If I find anything else, I'll let you know.

Best regards and thanks for sharing your code!

Ege

**Comment by lorenlugosch** [comment]
Ah, thanks Ege! I'll add this fix later. (Sorry, just saw this.)

**Comment by lorenlugosch** [comment]
Done!

**Comment by david-waterworth** [comment]
@lorenlugosch this issue also (appears to) cause a crash in the models.py implementation

https://github.com/lorenlugosch/pytorch_HMM/blob/40859fdf356ed7fd63913af704f351d81201a099/models.py#L140

I removed `genbmm` and replaced with the codes of @egaznep and it works - although I'm not sure if it's the same issue - it's hard to debug, the GPU version just throws an illegal memory access error and genbmm doesn't work on CPU which is usually easier to debug.

Also according to https://discuss.pytorch.org/t/solved-pytorch1-5-runtimeerror-one-of-the-variables-needed-for-gradient-computation-has-been-modified-by-an-inplace-operation/90256 it seems that "Before 1.5, these tests were not working properly for the optimizers. That’s why you didn’t see any error. But the computed gradients were not correct."
